src/wallet/rpcnames.cpp

```diff
@@ -524,13 +664,21 @@ name_firstupdate ()
   }
 
   const CNameScript prevNameOp(prevOut.scriptPubKey);
+
+  if (!fixedRand)
+    {
+      bool saltOK = getNameSalt (pwallet, name, prevOut.scriptPubKey, rand);
+      if (!saltOK)
+          throw JSONRPCError (RPC_TRANSACTION_ERROR, "could not generate rand for txid");
+      if (!saltMatchesHash (name, rand, prevNameOp.getOpHash ()))
```

Comment:
Since we have another check below already, I think we can remove this here; what do you think?  The only difference is that the error message here is slightly more explicit in this case.

Reply:
No, because if `!saltOK`, `rand` was never written to. If that happens, `rand` is still uninitialized. I don't think it's guaranteed that `std::vector<unsigned char>` will always be initialized to zero, which means you either get UB, or the compiler can constant-fold it to `if (true)` by deciding that `rand` initialized to whatever happens to be the value of `expectedSalt`.

Even if `std::vector` is guaranteed to be zero-initialized, it feels a bit uncomfortable to rely on this behavior.

Now, you could check `if (!saltOK || ...)`, but at that point you might as well do two checks.

Reply:
I meant to leave the `!saltOK` check here, but remove the `saltMatchesHash` one (sorry if that was unclear).  So if `!saltOK`, it will have thrown anyway.  But if the salt is OK and just doesn't match the name, then the check just a few lines below and outside of the `if` is guaranteed to fail (as it does basically the same thing).

Reply:
Yeah, I just think the error message is slightly clearer this way. "generated rand for txid does not match" => that's not a deterministic salt, "rand value is wrong" => what? there's no rand value here

Same reason as for why I throw "scan for previous txid failed" instead of just letting it get to "could not generate rand for txid" or, failing that, "previous txid not found". The old error messages are slightly confusing with the new API.